torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -153,6 +153,32 @@ void initializeStreamsEvents(
   }
 }
 
+void initializeStreamsEvents(
+    std::vector<std::vector<at::Tensor>>& outputs,
+    std::vector<std::vector<at::cuda::CUDAStream>>& streams,
+    std::vector<std::vector<at::cuda::CUDAEvent>>& events) {
+  at::cuda::OptionalCUDAGuard guard;
+  streams.resize(outputs.size());
+  events.resize(outputs.size());
+
+  for (size_t i = 0; i < outputs.size(); i++) {
+    streams[i].reserve(outputs[i].size());
+    events[i].resize(outputs[i].size());
+    for (size_t j = 0; j < outputs[i].size(); j++) {
+      guard.set_index(outputs[i][j].get_device());
+      // Record event on current stream
+      events[i][j].record(at::cuda::getCurrentCUDAStream());
+      // Get a non-default stream to execute asynchronous CUDA operations
+      // on for these outputs. This ensures that the default stream used
+      // by the caller is not occupied by c10d related operations.
+      streams[i].push_back(at::cuda::getStreamFromPool(
+          /* isHighPriority */ true, outputs[i][j].get_device()));
+      // Ensure the new stream is synchronized with the current stream.
+      events[i][j].block(streams[i][j]);
+    }
+  }
+}
+
```

Comment:
As discussed offline, we can probably allocate one stream per unique device index as well as just a single event per unique device index. There can only be 1 "current" stream per device, so as long as we synchronize with each one of them we're fine. CUDAStream is copyable so it can still be a nested vector of streams (it will just contain copies). CUDAEvent is NOT copyable, so it would require some other structure. We could do something like indirection where you have only a vector of streams and vector of events and then a nested vector of ints that correspond to the tensors, that index into the per-device streams and events vectors.